File: kernel/drivers/pci/pci.c
#include "pci.h"
#include "klog.h"
#include <stddef.h>

/* PCI config-space access via the standard I/O ports.
 *
 * CONFIG_ADDRESS layout (bit 31 set = enabled):
 *   bits 31:31  enable
 *   bits 23:16  bus number
 *   bits 15:11  device number
 *   bits 10:8   function number
 *   bits 7:2    register offset (dword aligned)
 * Writes go to CONFIG_DATA. */

#define PCI_CONFIG_ADDRESS 0xCF8u
#define PCI_CONFIG_DATA    0xCFCu

#define PCI_VENDOR_ID_OFFSET   0x00u
#define PCI_HEADER_TYPE_OFFSET 0x0Eu

#define PCI_VENDOR_NONE 0xFFFFu

static inline uint32_t inl(uint16_t port) {
    uint32_t val;
    __asm__ volatile ("inl %1, %0" : "=a"(val) : "Nd"(port));
    return val;
}

static inline void outl(uint16_t port, uint32_t val) {
    __asm__ volatile ("outl %0, %1" : : "a"(val), "Nd"(port));
}

static uint32_t pci_make_address(uint8_t bus, uint8_t dev, uint8_t func, uint8_t offset) {
    return 0x80000000u
         | ((uint32_t)bus  << 16)
         | ((uint32_t)dev  << 11)
         | ((uint32_t)func << 8)
         | (offset & 0xFCu);
}

uint32_t pci_config_read32(uint8_t bus, uint8_t dev, uint8_t func, uint8_t offset) {
    outl(PCI_CONFIG_ADDRESS, pci_make_address(bus, dev, func, offset));
    return inl(PCI_CONFIG_DATA);
}

void pci_config_write32(uint8_t bus, uint8_t dev, uint8_t func, uint8_t offset, uint32_t value) {
    outl(PCI_CONFIG_ADDRESS, pci_make_address(bus, dev, func, offset));
    outl(PCI_CONFIG_DATA, value);
}

uint16_t pci_config_read16(uint8_t bus, uint8_t dev, uint8_t func, uint8_t offset) {
    return (uint16_t)(pci_config_read32(bus, dev, func, offset) >> ((offset & 3u) * 8u));
}

uint8_t pci_config_read8(uint8_t bus, uint8_t dev, uint8_t func, uint8_t offset) {
    return (uint8_t)(pci_config_read32(bus, dev, func, offset) >> ((offset & 3u) * 8u));
}

/* ---- device table ---- */

static pci_device_t pci_devices[PCI_MAX_DEVICES];
static uint32_t     pci_count = 0;

uint32_t pci_device_count(void) {
    return pci_count;
}

const pci_device_t *pci_device_at(uint32_t i) {
    if (i >= pci_count) return NULL;
    return &pci_devices[i];
}

int pci_find_device(uint16_t vendor, uint16_t device_id) {
    for (uint32_t i = 0; i < pci_count; i++) {
        if (pci_devices[i].vendor == vendor &&
            pci_devices[i].device_id == device_id) {
            return (int)i;
        }
    }
    return -1;
}

uint8_t pci_cap_first(uint8_t bus, uint8_t dev, uint8_t func) {
    /* Bit 4 of the status register: capabilities list present. */
    uint16_t status = pci_config_read16(bus, dev, func, 0x06u);
    if (!(status & 0x0010u)) return 0;

    uint8_t ptr = pci_config_read8(bus, dev, func, 0x34u);
    /* Valid capability pointers live in 0x40-0xFF. */
    return (ptr >= 0x40u) ? ptr : 0;
}

uint8_t pci_cap_next(uint8_t bus, uint8_t dev, uint8_t func, uint8_t offset) {
    if (offset == 0 || offset >= 0xFFu) return 0;
    uint8_t next = pci_config_read8(bus, dev, func, (uint8_t)(offset + 1));
    return (next >= 0x40u) ? next : 0;
}

/* Records one present function, guarding the table bounds. */
static void pci_record(uint8_t bus, uint8_t dev, uint8_t func) {
    if (pci_count >= PCI_MAX_DEVICES) {
        klog("PCI: device table full (%u), skipping bus %u dev %u func %u\n",
             PCI_MAX_DEVICES, bus, dev, func);
        return;
    }

    pci_device_t *p = &pci_devices[pci_count];
    p->bus = bus;
    p->dev = dev;
    p->func = func;
    p->vendor     = pci_config_read16(bus, dev, func, 0x00u);
    p->device_id  = pci_config_read16(bus, dev, func, 0x02u);
    p->revision   = pci_config_read8 (bus, dev, func, 0x08u);
    p->prog_if    = pci_config_read8 (bus, dev, func, 0x09u);
    p->sub_class  = pci_config_read8 (bus, dev, func, 0x0Au);
    p->base_class = pci_config_read8 (bus, dev, func, 0x0Bu);
    p->header_type = (uint8_t)(pci_config_read8(bus, dev, func, PCI_HEADER_TYPE_OFFSET) & 0x7Fu);

    /* Base address registers only exist in header type 0. */
    if (p->header_type == 0x00u) {
        for (int i = 0; i < 6; i++) {
            p->bars[i] = pci_config_read32(bus, dev, func, (uint8_t)(0x10u + i * 4u));
        }
    }

    pci_count++;
}

static const char *pci_class_name(uint8_t base_class) {
    switch (base_class) {
        case 0x00: return "Legacy";
        case 0x01: return "Storage";
        case 0x02: return "Network";
        case 0x03: return "Display";
        case 0x04: return "Multimedia";
        case 0x05: return "Memory";
        case 0x06: return "Bridge";
        case 0x07: return "Comm";
        case 0x08: return "System peripheral";
        case 0x09: return "Input";
        case 0x0C: return "Serial bus";       /* USB / SMBus */
        case 0x0D: return "Wireless";
        case 0x0E: return "Intelligent I/O";
        case 0xFF: return "VGA/unknown";
        default:   return "Other";
    }
}

uint32_t pci_init(void) {
    pci_count = 0;

    for (uint32_t bus = 0; bus < 256; bus++) {
        for (uint32_t dev = 0; dev < 32; dev++) {
            uint16_t vendor = pci_config_read16((uint8_t)bus, (uint8_t)dev, 0, PCI_VENDOR_ID_OFFSET);
            if (vendor == PCI_VENDOR_NONE) continue;

            uint8_t header = pci_config_read8((uint8_t)bus, (uint8_t)dev, 0, PCI_HEADER_TYPE_OFFSET);
            uint32_t num_funcs = (header & 0x80u) ? 8u : 1u;

            for (uint32_t func = 0; func < num_funcs; func++) {
                vendor = pci_config_read16((uint8_t)bus, (uint8_t)dev, (uint8_t)func, PCI_VENDOR_ID_OFFSET);
                if (vendor == PCI_VENDOR_NONE) continue;
                pci_record((uint8_t)bus, (uint8_t)dev, (uint8_t)func);
            }
        }
    }

    klog("PCI: %u device(s) found\n", pci_count);
    for (uint32_t i = 0; i < pci_count; i++) {
        const pci_device_t *p = &pci_devices[i];
        klog("  [%x:%x.%u] %x:%x class=%x/%x %s\n",
             p->bus, p->dev, p->func, p->vendor, p->device_id,
             p->base_class, p->sub_class,
             pci_class_name(p->base_class));
    }

    return pci_count;
}

File: rust/src/drivers/pci.rs
//! Minimal PCI bus enumeration over the 0xCF8/0xCFC config ports.
//! Faithful port of `kernel/drivers/pci/pci.c` (C).
//!
//! The scan walks buses 0-255, devices 0-31, and functions 0-7
//! (skipping function 1-7 of single-function devices via the
//! multifunction bit in the header type). Every present function is
//! recorded in a fixed-size, bounds-checked table. Only the surface
//! the VirtIO input driver needs is ported: device discovery, config
//! accessors, and capability-list walking.

pub const PCI_MAX_DEVICES: usize = 256;

const PCI_CONFIG_ADDRESS: u16 = 0xCF8;
const PCI_CONFIG_DATA: u16 = 0xCFC;

const PCI_VENDOR_ID_OFFSET: u8 = 0x00;
const PCI_HEADER_TYPE_OFFSET: u8 = 0x0E;

const PCI_VENDOR_NONE: u16 = 0xFFFF;

#[derive(Clone, Copy)]
pub struct PciDevice {
    pub bus: u8,
    pub dev: u8,
    pub func: u8,
    pub vendor: u16,
    pub device_id: u16,
    pub revision: u8,
    pub prog_if: u8,
    pub sub_class: u8,
    pub base_class: u8,
    pub header_type: u8,
    pub bars: [u32; 6],
}

const PCI_DEVICE_ZERO: PciDevice = PciDevice {
    bus: 0,
    dev: 0,
    func: 0,
    vendor: 0,
    device_id: 0,
    revision: 0,
    prog_if: 0,
    sub_class: 0,
    base_class: 0,
    header_type: 0,
    bars: [0; 6],
};

static mut PCI_DEVICES: [PciDevice; PCI_MAX_DEVICES] = [PCI_DEVICE_ZERO; PCI_MAX_DEVICES];
static mut PCI_COUNT: usize = 0;

#[inline]
unsafe fn inl(port: u16) -> u32 {
    let val: u32;
    unsafe {
        core::arch::asm!(
            "in eax, dx",
            out("eax") val,
            in("dx") port,
            options(nomem, nostack, preserves_flags)
        );
    }
    val
}

#[inline]
unsafe fn outl(port: u16, val: u32) {
    unsafe {
        core::arch::asm!(
            "out dx, eax",
            in("dx") port,
            in("eax") val,
            options(nomem, nostack, preserves_flags)
        );
    }
}

/* CONFIG_ADDRESS layout (bit 31 set = enabled): bits 23:16 bus,
 * 15:11 device, 10:8 function, 7:2 register offset (dword aligned). */
fn make_address(bus: u8, dev: u8, func: u8, offset: u8) -> u32 {
    0x8000_0000
        | ((bus as u32) << 16)
        | ((dev as u32) << 11)
        | ((func as u32) << 8)
        | (offset as u32 & 0xFC)
}

pub fn config_read32(bus: u8, dev: u8, func: u8, offset: u8) -> u32 {
    unsafe {
        outl(PCI_CONFIG_ADDRESS, make_address(bus, dev, func, offset));
        inl(PCI_CONFIG_DATA)
    }
}

// Unused by the current drivers; kept as part of the ported C API
// surface (future drivers will need it for enabling bus mastering).
#[allow(dead_code)]
pub fn config_write32(bus: u8, dev: u8, func: u8, offset: u8, value: u32) {
    unsafe {
        outl(PCI_CONFIG_ADDRESS, make_address(bus, dev, func, offset));
        outl(PCI_CONFIG_DATA, value);
    }
}

pub fn config_read16(bus: u8, dev: u8, func: u8, offset: u8) -> u16 {
    (config_read32(bus, dev, func, offset) >> ((offset & 3) * 8)) as u16
}

pub fn config_read8(bus: u8, dev: u8, func: u8, offset: u8) -> u8 {
    (config_read32(bus, dev, func, offset) >> ((offset & 3) * 8)) as u8
}

/// Number of devices recorded by `init`.
pub fn device_count() -> usize {
    unsafe { PCI_COUNT }
}

/// Device at table index `i`, or None if out of range.
pub fn device_at(i: usize) -> Option<&'static PciDevice> {
    if i >= unsafe { PCI_COUNT } {
        return None;
    }
    Some(unsafe { &PCI_DEVICES[i] })
}

/// Table index of the first device matching (vendor, device_id), or
/// None.
// Unused by the current drivers; kept as part of the ported C API
// surface.
#[allow(dead_code)]
pub fn find_device(vendor: u16, device_id: u16) -> Option<usize> {
    let count = unsafe { PCI_COUNT };
    for i in 0..count {
        let d = unsafe { &PCI_DEVICES[i] };
        if d.vendor == vendor && d.device_id == device_id {
            return Some(i);
        }
    }
    None
}

/// Config-space offset of the first capability (0 if none). Valid
/// capability pointers live in 0x40-0xFF.
pub fn cap_first(bus: u8, dev: u8, func: u8) -> u8 {
    let status = config_read16(bus, dev, func, 0x06);
    if status & 0x0010 == 0 {
        return 0; /* bit 4 of status: capabilities list present */
    }
    let ptr = config_read8(bus, dev, func, 0x34);
    if ptr >= 0x40 {
        ptr
    } else {
        0
    }
}

/// Capability after `offset` (0 if none).
pub fn cap_next(bus: u8, dev: u8, func: u8, offset: u8) -> u8 {
    if offset == 0 || offset >= 0xFF {
        return 0;
    }
    let next = config_read8(bus, dev, func, offset + 1);
    if next >= 0x40 {
        next
    } else {
        0
    }
}

fn record(bus: u8, dev: u8, func: u8) {
    let count = unsafe { PCI_COUNT };
    if count >= PCI_MAX_DEVICES {
        crate::kprintln!(
            "[rust] PCI: device table full ({}), skipping bus {} dev {} func {}",
            PCI_MAX_DEVICES,
            bus,
            dev,
            func
        );
        return;
    }
    unsafe {
        let p = &mut PCI_DEVICES[count];
        p.bus = bus;
        p.dev = dev;
        p.func = func;
        p.vendor = config_read16(bus, dev, func, 0x00);
        p.device_id = config_read16(bus, dev, func, 0x02);
        p.revision = config_read8(bus, dev, func, 0x08);
        p.prog_if = config_read8(bus, dev, func, 0x09);
        p.sub_class = config_read8(bus, dev, func, 0x0A);
        p.base_class = config_read8(bus, dev, func, 0x0B);
        p.header_type = config_read8(bus, dev, func, PCI_HEADER_TYPE_OFFSET) & 0x7F;

        /* Base address registers only exist in header type 0. */
        if p.header_type == 0x00 {
            for i in 0..6 {
                p.bars[i] = config_read32(bus, dev, func, 0x10 + (i as u8) * 4);
            }
        }

        PCI_COUNT += 1;
    }
}

fn class_name(base_class: u8) -> &'static str {
    match base_class {
        0x00 => "Legacy",
        0x01 => "Storage",
        0x02 => "Network",
        0x03 => "Display",
        0x04 => "Multimedia",
        0x05 => "Memory",
        0x06 => "Bridge",
        0x07 => "Comm",
        0x08 => "System peripheral",
        0x09 => "Input",
        0x0C => "Serial bus",
        0x0D => "Wireless",
        0x0E => "Intelligent I/O",
        0xFF => "VGA/unknown",
        _ => "Other",
    }
}

/// Scans all buses and fills the device table. Returns the number of
/// devices found (never more than PCI_MAX_DEVICES).
pub fn init() -> usize {
    unsafe { PCI_COUNT = 0 };

    for bus in 0..256u32 {
        for dev in 0..32u32 {
            let mut vendor = config_read16(bus as u8, dev as u8, 0, PCI_VENDOR_ID_OFFSET);
            if vendor == PCI_VENDOR_NONE {
                continue;
            }
            let header = config_read8(bus as u8, dev as u8, 0, PCI_HEADER_TYPE_OFFSET);
            let num_funcs = if header & 0x80 != 0 { 8 } else { 1 };
            for func in 0..num_funcs {
                vendor = config_read16(bus as u8, dev as u8, func as u8, PCI_VENDOR_ID_OFFSET);
                if vendor == PCI_VENDOR_NONE {
                    continue;
                }
                record(bus as u8, dev as u8, func as u8);
            }
        }
    }

    let count = unsafe { PCI_COUNT };
    crate::kprintln!("[rust] PCI: {} device(s) found", count);
    for i in 0..count {
        let d = unsafe { &PCI_DEVICES[i] };
        crate::kprintln!(
            "[rust]   [{:x}:{:x}.{}] {:x}:{:x} class={:x}/{:x} {}",
            d.bus,
            d.dev,
            d.func,
            d.vendor,
            d.device_id,
            d.base_class,
            d.sub_class,
            class_name(d.base_class)
        );
    }
    count
}
